File: 04 Instrumental Variables in Action/Figure 4-6-1.py
#!/usr/bin/env python
"""
Create Figure 4-6-1 in MHE
Tested on Python 3.4
"""

import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.sandbox.regression import gmm
import matplotlib.pyplot as plt
import random
import math
from scipy.linalg import eigh

# Number of simulations
nsims = 10

# Set seed
random.seed(461)

# Set parameters
Sigma   = [[1.0, 0.8],
           [0.8, 1.0]]
mu      = [0, 0]
errors  = np.random.multivariate_normal(mu, Sigma, 1000)
eta     = errors[:, 0]
xi      = errors[:, 1]

# Create Z, x, y
Z  = np.random.multivariate_normal([0] * 20, np.identity(20), 1000)
x  = 0.1 * Z[: , 0] + xi
y  = x + eta
x  = sm.add_constant(x)
Z  = sm.add_constant(x)

ols  = sm.OLS(y, x).fit().params[1]
# tsls = np.linalg.inv(np.transpose(Z).dot(x)).dot(np.transpose(Z).dot(y))[1]
tsls = gmm.IV2SLS(y, x, Z).fit().params[1]

def LIML(exogenous, endogenous, instruments):
    y = exogenous
    x = endogenous
    Z = instruments
    I = np.eye(y.shape[0])
    Mz = I - Z.dot(np.linalg.inv(np.transpose(Z).dot(Z))).dot(np.transpose(Z))
    Mx = I - x.dot(np.linalg.inv(np.transpose(x).dot(x))).dot(np.transpose(x))
    A = np.transpose(np.hstack((y, x[:,1]))).dot(Mz).dot(np.hstack((y, x[:,1])))
    k = 1 
    beta = np.linalg.inv(np.transpose(Z).dot(I - k * M).dot(Z)).dot(np.transpose(Z).dot(I - k * M)).dot(y)
    return 

# End of script

File: 04 Instrumental Variables in Action/Figure 4-6-1.jl
# Julia code for Table 4-6-1                #
# Required packages                         #
# - DataFrames: data manipulation / storage #
# - Distributions: extended stats functions #
# - FixedEffectModels: IV regression        #
using DataFrames
using Distributions
using FixedEffectModels
using GLM
using Gadfly

# Number of simulations
nsims = 1000

# Set seed
srand(113643)

# Set parameters
Sigma = [1.0 0.8;
         0.8 1.0]
N     = 1000

function irrelevantInstrMC()
    # Create Z, xi and eta
    Z      = DataFrame(transpose(rand(MvNormal(eye(20)), N)))
    errors = DataFrame(transpose(rand(MvNormal(Sigma), N)))

    # Rename columns of Z and errors
    names!(Z, [Symbol("z$i") for i in 1:20])
    names!(errors, [:eta, :xi])

    # Create y and x
    df     = hcat(Z, errors);
    df[:x] = 0.1 .* df[:z1] .+ df[:xi]
    df[:y] = df[:x] .+ df[:eta]

    # Run regressions
    ols  = coef(lm(@formula(y ~ x), df))[2]
    tsls = coef(reg(df, @model(y ~ z1  + z2  + z3  + z4  + z5  + z6  + z7  + z8  + z9  + z10 +
                             z11 + z12 + z13 + z14 + z15 + z16 + z17 + z18 + z19 + z20)))[2]
    return([ols tsls])
end

# Simulate IV regressions
simulation_results = zeros(nsims, 2);
for i = 1:nsims
    simulation_results[i, :] = irrelevantInstrMC()
end

# Create empirical CDFs from simulated results
ols_ecdf  = ecdf(simulation_results[:, 1])
tsls_ecdf = ecdf(simulation_results[:, 2])

# Plot the empirical CDFs of each estimator
p = plot(layer(ols_ecdf, 0, 2.5, Theme(default_color = colorant"red")),
         layer(tsls_ecdf, 0, 2.5, Theme(line_style = :dot)),
         layer(xintercept = [0.5], Geom.vline,
               Theme(default_color = colorant"black", line_style = :dot)),
         layer(yintercept = [0.5], Geom.hline,
               Theme(default_color = colorant"black", line_style = :dot)),
         Guide.xlabel("Estimated β"),
         Guide.ylabel("F<sub>n</sub>(Estimated β)"))

# Export figure as .png
draw(PNG("Figure 4-6-1-Julia.png", 7inch, 6inch), p)

# End of script
